### PR TITLE
fix(ha_nexus): cleanup older unregistered shutdown targets

### PR DESCRIPTION
**Pull request by abhilashshetty04**
This PR allows cluster agents to cleanup shutdown targets of the volume which are not registered as Subsystems in the Frontend node. Cleanup starts when we exhaust all nodes during target republish.

Signed-off-by: Abhilash Shetty <abhilash.shetty@datacore.com>

### CODE REVIEW

File: common/src/transport_api/v0.rs [hunk 1]
```diff
@@ -30,6 +30,8 @@ impl_message!(DestroyPool);
 impl_vector_request!(Pools, Pool);
 impl_message!(GetPools);
 
+impl_vector_request!(NvmeSubsystems, NvmeSubsystem);
```
Comment (avishnu) [inline common/src/transport_api/v0.rs:33]: Should this not be in v1.rs?

File: common/src/types/v0/transport/cluster_agent.rs [hunk 2]
```diff
@@ -83,6 +83,42 @@ impl ReportFailedPaths {
...
+        Self {
+            address: target_addr,
+        }
+    }
+    pub fn address(&self) -> String {
+        self.address.clone()
+    }
+}
+
+// impl From<NvmeSubsystem> for String {
```
Comment (avishnu) [inline common/src/types/v0/transport/cluster_agent.rs]: Is this commented block needed? If not, let's delete the same.

Comment (abhilashshetty04) [reply common/src/types/v0/transport/cluster_agent.rs]: Removed it now.

File: control-plane/agents/src/bin/core/controller/resources/operations.rs [hunk 3]
```diff
@@ -139,4 +139,6 @@ pub(crate) trait ResourceShutdownOperations {
         registry: &Registry,
         request: &Self::RemoveShutdownTargets,
     ) -> Result<(), SvcError>;
+
```
Comment (avishnu) [inline control-plane/agents/src/bin/core/controller/resources/operations.rs]: Is this needed?

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 4]
```diff
@@ -563,22 +564,29 @@ impl ResourceShutdownOperations for OperationGuardArc<VolumeSpec> {
     ) -> Result<(), SvcError> {
         let shutdown_nexuses = registry
             .specs()
-            .get_volume_shutdown_nexuses(&request.uuid)
+            .get_volume_shutdown_nexuses(&request.uuid())
```
Comment (avishnu) [inline control-plane/agents/src/bin/core/volume/operations.rs]: How has this been working till now?

File: common/src/transport_api/v0.rs (hunk 1 above)
Comment (abhilashshetty04) [reply common/src/transport_api/v0.rs:33]: NvmeSubsystem/s  are both control plane  specific structs. This macro allows us to have vector representation of **NvmeSubsystem** struct. Vector object would be **NvmeSubsystems** .

We have saparate v0, v1 specs when grpc is between control plane and data plane only.

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 4 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/volume/operations.rs]: uuid field in the DestroyShutdownTargets used to be public field. Now i have added a getter fn uuid(). Made the field private.

File: control-plane/agents/src/bin/core/controller/resources/operations.rs (hunk 3 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/resources/operations.rs]: Not needed. Removed it

File: control-plane/agents/src/bin/ha/node/server.rs [hunk 5]
```diff
@@ -273,6 +301,26 @@ fn parse_uri(new_path: &str) -> Result<ParsedUri, SvcError> {
     ParsedUri::new(uri)
 }
 
+/// Gets all Nvme subsystems registered for a given volume.
+fn get_subsystems(parsed_uri: &ParsedUri) -> Result<Vec<NvmeCtrller>, SvcError> {
```
Comment (abhilashshetty04) [inline control-plane/agents/src/bin/ha/node/server.rs]: @tiagolobocastro , I think this has to go into mayastor-dependency. Added this here as of now as i dont have merge access. Will raise new PR in that repo.

File: common/src/types/v0/transport/cluster_agent.rs [hunk 6]
```diff
@@ -83,6 +83,43 @@ impl ReportFailedPaths {
     }
 }
 
+#[derive(Debug)]
+/// Request struct to get Nvme subsystems registered for a given nqn.
```
Comment (tiagolobocastro) [inline common/src/types/v0/transport/cluster_agent.rs:86]: nit: swap 86 with 87

File: common/src/types/v0/transport/cluster_agent.rs [hunk 7]
```diff
@@ -83,6 +83,43 @@ impl ReportFailedPaths {
...
+
+impl NvmeSubsystem {
+    /// Creates an instance of this struct. address is the IP address of the Nvme Subsystem.
+    pub fn new(target_addr: String) -> Self {
+        Self {
+            address: target_addr,
+        }
+    }
+    /// Get IP address of the Nvme Subsystem.
+    pub fn address(&self) -> String {
```
Comment (tiagolobocastro) [inline common/src/types/v0/transport/cluster_agent.rs]: nit: consider returning &str instead of String and let caller to_string if needed.

File: common/src/types/v0/transport/volume.rs [hunk 8]
```diff
@@ -605,12 +605,25 @@ impl From<&ReplicaTopology> for models::ReplicaTopology {
...
-    pub fn new(uuid: VolumeId) -> Self {
-        DestroyShutdownTargets { uuid }
+    pub fn new(uuid: VolumeId, registered_targets: Option<Vec<String>>) -> Self {
+        DestroyShutdownTargets {
+            uuid,
+            registered_targets,
+        }
+    }
+    /// Get volumeId.
+    pub fn uuid(&self) -> VolumeId {
```
Comment (tiagolobocastro) [inline common/src/types/v0/transport/volume.rs]: consider returning reference instead

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 9]
```diff
@@ -195,9 +195,63 @@ impl SwitchOverRequest {
             {
                 error!(volume.uuid=%self.volume_id, %error, "Cancelling switchover");
                 self.stage = Stage::Errored;
-                Err(error)
+                Err(anyhow!(error.to_string()))
             }
-            Err(error) => Err(error),
+            Err(err) if err.kind == ReplyErrorKind::ResourceExhausted => {
+                // Cleanup volume targets that are not registered as Nvme Subsystems in the Node.
+                if self.new_path.is_some() {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs]: if let Some(new_path) = self.new_path.clone() { ... }

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 10]
```diff
@@ -195,9 +195,63 @@ impl SwitchOverRequest {
...
+                    if let Ok(uri) = Uri::builder()
+                        .scheme("http")
+                        .authority(self.callback_uri.to_string())
+                        .path_and_query("")
+                        .build()
+                    {
+                        let node_client = NodeAgentClient::new(uri, None).await;
+                        let request = GetController::new(self.new_path.clone().unwrap());
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs]: new_path from if let Some(...)

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 11]
```diff
@@ -195,9 +195,63 @@ impl SwitchOverRequest {
...
+                                return Ok(());
+                            }
+                            _ => {
+                                let err = format!("Failed to list Nvme subsystems for the volume");
+                                Err(anyhow!(err))
+                            }
+                        }
+                    } else {
+                        self.stage = Stage::Errored;
+                        Err(anyhow!("Failed to get node agent grpc endpoint"))
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs]: Failed to get or ill-formed ?

File: control-plane/agents/src/bin/ha/node/server.rs (hunk 5 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/ha/node/server.rs]: IIRC @Abhinandan-Purkait already added this functionality, or similar?

File: control-plane/agents/src/common/errors.rs [hunk 12]
```diff
@@ -231,6 +231,8 @@ pub enum SvcError {
     NvmeParseError {},
     #[snafu(display("Nvme connect failed : {}", details))]
     NvmeConnectError { details: String },
+    #[snafu(display("No Subsystems found for {}", nvme_path))]
+    NoSubsystemFound { nvme_path: String },
```
Comment (tiagolobocastro) [inline control-plane/agents/src/common/errors.rs]: Already exists above: SubsystemNotFound

File: control-plane/grpc/proto/v1/ha/node_agent.proto [hunk 13]
```diff
@@ -21,4 +22,22 @@ message ReplacePathRequest {
 
   // Pubish context of the volume
   common.MapWrapper publish_context = 3;
+}
+
+// Get Controller request message.
+message GetNvmeControllerRequest{
```
Comment (tiagolobocastro) [inline control-plane/grpc/proto/v1/ha/node_agent.proto]: add space before { on all messages

File: control-plane/grpc/proto/v1/ha/node_agent.proto [hunk 14]
```diff
@@ -21,4 +22,22 @@ message ReplacePathRequest {
...
+  string nvme_path = 1;
```
Comment (tiagolobocastro) [inline control-plane/grpc/proto/v1/ha/node_agent.proto:30]: add comment above each field

File: control-plane/grpc/src/operations/ha_node/traits.rs [hunk 15]
```diff
@@ -22,6 +23,60 @@ pub trait NodeAgentOperations: Send + Sync {
...
+        }
+    }
+}
+
+impl From<NvmeControllers> for NvmeSubsystems {
+    fn from(grpc_type_subsys: NvmeControllers) -> Self {
+        let mut controllers: Vec<NvmeSubsystem> = vec![];
+        for controller in grpc_type_subsys.target_address {
+            controllers.push(NvmeSubsystem::new(controller));
+        }
```
Comment (tiagolobocastro) [inline control-plane/grpc/src/operations/ha_node/traits.rs]:
```suggestion
        let controllers = grpc_type_subsys.target_address.into_iter().map(NvmeSubsystem::new).collect();
```

File: control-plane/agents/src/bin/ha/node/server.rs (hunk 5 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/node/server.rs]: Its slightly different, this method will return list(if applicable) of NvmeSubsystems based on nqn. Whereas what @Abhinandan-Purkait has added will filter using address and nqn and return the First Subsystem that matches.

File: common/src/types/v0/transport/cluster_agent.rs (hunk 6 above)
Comment (abhilashshetty04) [reply common/src/types/v0/transport/cluster_agent.rs:86]: Done

File: control-plane/agents/src/bin/ha/cluster/switchover.rs (hunk 9 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs]: Done

File: control-plane/agents/src/bin/ha/cluster/switchover.rs (hunk 10 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs]: Incorporated

File: control-plane/agents/src/bin/ha/cluster/switchover.rs (hunk 11 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs]: Incorporated

File: control-plane/agents/src/common/errors.rs (hunk 12 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/common/errors.rs]: Incorporated

File: control-plane/grpc/proto/v1/ha/node_agent.proto (hunk 14 above)
Comment (abhilashshetty04) [reply control-plane/grpc/proto/v1/ha/node_agent.proto:30]: Incorporated

File: control-plane/grpc/src/operations/ha_node/traits.rs (hunk 15 above)
Comment (abhilashshetty04) [reply control-plane/grpc/src/operations/ha_node/traits.rs]: Thanks. Incorporated

File: common/src/types/v0/transport/cluster_agent.rs (hunk 7 above)
Comment (abhilashshetty04) [reply common/src/types/v0/transport/cluster_agent.rs]: Incorporated

File: control-plane/grpc/proto/v1/ha/node_agent.proto (hunk 13 above)
Comment (abhilashshetty04) [reply control-plane/grpc/proto/v1/ha/node_agent.proto]: Incorporated

Comment (abhilashshetty04) [reply control-plane/grpc/proto/v1/ha/node_agent.proto]: Incorporated

File: control-plane/grpc/proto/v1/ha/node_agent.proto (hunk 14 above)
Comment (abhilashshetty04) [reply control-plane/grpc/proto/v1/ha/node_agent.proto:30]: Incorporated

File: common/src/types/v0/transport/volume.rs (hunk 8 above)
Comment (abhilashshetty04) [reply common/src/types/v0/transport/volume.rs]: Thanks. Incorporated

File: common/src/types/v0/transport/cluster_agent.rs [hunk 16]
```diff
@@ -83,6 +83,43 @@ impl ReportFailedPaths {
...
 }
 
+/// Request struct to get Nvme subsystems registered for a given nqn.
+#[derive(Debug)]
+pub struct GetController {
+    nvme_path: String,
+}
+
+impl GetController {
+    /// Constructor to create instance of the struct. nvme_path is the device uri of the target
```
Comment (Abhinandan-Purkait) [inline common/src/types/v0/transport/cluster_agent.rs]: `struct.` -> `struct,`

File: common/src/types/v0/transport/cluster_agent.rs [hunk 17]
```diff
@@ -83,6 +83,43 @@ impl ReportFailedPaths {
...
+    /// associated with volume.
+    pub fn new(nvme_path: String) -> Self {
```
Comment (Abhinandan-Purkait) [inline common/src/types/v0/transport/cluster_agent.rs:95]: `associated` -> `Associated`

File: common/src/types/v0/transport/cluster_agent.rs [hunk 18]
```diff
@@ -83,6 +83,43 @@ impl ReportFailedPaths {
...
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+/// Represents targets address of Nvme Subsystem.
+pub struct NvmeSubsystem {
+    address: String,
+}
+
+impl NvmeSubsystem {
+    /// Creates an instance of this struct. address is the IP address of the Nvme Subsystem.
```
Comment (Abhinandan-Purkait) [inline common/src/types/v0/transport/cluster_agent.rs]: struct. -> struct,

**Review by Abhinandan-Purkait — APPROVED** [review]

File: common/src/types/v0/transport/cluster_agent.rs (hunk 17 above)
Comment (abhilashshetty04) [reply common/src/types/v0/transport/cluster_agent.rs:95]: I have added nvme path documentation in struct defination itself.

File: common/src/types/v0/transport/cluster_agent.rs (hunk 18 above)
Comment (abhilashshetty04) [reply common/src/types/v0/transport/cluster_agent.rs]: Incorporated

File: common/src/types/v0/transport/cluster_agent.rs (hunk 16 above)
Comment (abhilashshetty04) [reply common/src/types/v0/transport/cluster_agent.rs]: Incorporated

File: control-plane/agents/src/bin/ha/node/server.rs (hunk 5 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/node/server.rs]: @tiagolobocastro . I have raised a separate PR for moving this function to mayastor-dependency repo.

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 19]
```diff
@@ -563,22 +564,29 @@ impl ResourceShutdownOperations for OperationGuardArc<VolumeSpec> {
...
-
         let mut result = Ok(());
+        let filter_needed = request.registered_targets().is_some();
         for nexus_res in shutdown_nexuses {
             match nexus_res.operation_guard_wait().await {
                 Ok(mut guard) => {
+                    let uuid = nexus_res.lock().clone().uuid;
+                    let nexus = registry.get_nexus(&uuid).await?;
+                    if filter_needed
+                        && filter_targets(request.registered_targets().unwrap_or_default(), nexus)?
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: nit: why not have `filter_targets` receive an `Option` by itself, then we can avoid having `filter_needed` and `unwrap_or_default` at the caller site?

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 20]
```diff
@@ -611,6 +619,20 @@ impl ResourceShutdownOperations for OperationGuardArc<VolumeSpec> {
     }
 }
 
+/// Checks if Nexus is present in registered target list. Returns true if yes.
+fn filter_targets(registered_target: Vec<String>, nexus: Nexus) -> Result<bool, SvcError> {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: this name is a big vague, how about `target_registered` or something like that?

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 21]
```diff
@@ -195,9 +195,66 @@ impl SwitchOverRequest {
             {
                 error!(volume.uuid=%self.volume_id, %error, "Cancelling switchover");
                 self.stage = Stage::Errored;
-                Err(error)
+                Err(anyhow!(error.to_string()))
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs]: Why is this change needed?

File: control-plane/agents/src/bin/ha/node/server.rs [hunk 22]
```diff
@@ -262,6 +268,36 @@ impl NodeAgentOperations for NodeAgentSvc {
...
+        match Subsystem::try_from_nqn(nqn.as_str()) {
+            Ok(subsys_list) => {
+                let mut controller_list = vec![];
+                for controller in subsys_list {
+                    if controller.state == "live" && controller.state == "connecting" {
+                        controller_list.push(NvmeCtrller::new(controller.clone().address))
+                    } else {
+                        continue;
+                    }
+                }
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/ha/node/server.rs]: nit: can replace this with something like:
```suggestion
                let controller_list = subsys_list
                                  .into_iter()
                                  .filter(|controller| controller.state == "live" && controller.state == "connecting")
                                  .map(|controller| NvmeCtrller::new(controller.address))
                                  .collect();
```

File: control-plane/grpc/proto/v1/ha/node_agent.proto [hunk 23]
```diff
@@ -21,4 +22,26 @@ message ReplacePathRequest {
...
+    // List of target address registered as Nvme Subsystem.
+    NvmeControllers nvme_controllers = 1;
+    // Error incase No subsystem found.
+    common.ReplyError error = 2;
+  }
+}
+
+// Message to represent target's address in the rpc response.
+message NvmeControllers {
+ //  Target address registered as Nvme Subsystem.
```
Comment (tiagolobocastro) [inline control-plane/grpc/proto/v1/ha/node_agent.proto]:
```suggestion
 // Target address registered as Nvme Subsystem.
```

File: control-plane/grpc/proto/v1/volume/volume.proto [hunk 24]
```diff
@@ -321,6 +321,11 @@ message ProbeResponse {
 message DestroyShutdownTargetRequest {
   // uuid of the volume
   google.protobuf.StringValue volume_id = 1;
+  optional RegisteredTargets registered_targets = 2;
```
Comment (tiagolobocastro) [inline control-plane/grpc/proto/v1/volume/volume.proto:325]: missing comments here and below

File: control-plane/grpc/src/operations/ha_node/traits.rs [hunk 25]
```diff
@@ -22,6 +23,61 @@ pub trait NodeAgentOperations: Send + Sync {
...
+    }
+}
+
+impl From<NvmeControllers> for NvmeSubsystems {
+    fn from(grpc_type_subsys: NvmeControllers) -> Self {
+        let controllers = grpc_type_subsys
+            .target_address
+            .into_iter()
+            .map(NvmeSubsystem::new)
```
Comment (tiagolobocastro) [inline control-plane/grpc/src/operations/ha_node/traits.rs:77]: 👍

File: control-plane/grpc/proto/v1/ha/node_agent.proto [hunk 26]
```diff
@@ -21,4 +22,26 @@ message ReplacePathRequest {
...
+  string nvme_path = 1;
+}
+
+// Get Controller response message.
+message GetNvmeControllerResponse {
+  oneof reply {
+    // List of target address registered as Nvme Subsystem.
+    NvmeControllers nvme_controllers = 1;
+    // Error incase No subsystem found.
+    common.ReplyError error = 2;
```
Comment (tiagolobocastro) [inline control-plane/grpc/proto/v1/ha/node_agent.proto:39]: q: do we need a ReplyError error here, is the normal grpc status not enough?

File: control-plane/agents/src/bin/ha/cluster/switchover.rs (hunk 21 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs]: In the same arm i had errors returned using anyhow! formatting if unable to get nvme_path and uri formatting. I was getting error that match arm had incompatible  return types.

File: control-plane/grpc/proto/v1/ha/node_agent.proto (hunk 26 above)
Comment (abhilashshetty04) [reply control-plane/grpc/proto/v1/ha/node_agent.proto:39]: We are sending ReplyError::not_found from node agent incase we didnt find any subsystem for nqn or any other error. In cluster agent, We match if its ReplyError::not_found . If yes, We cleanup all shutdown targets for the volume.

File: control-plane/grpc/proto/v1/volume/volume.proto (hunk 24 above)
Comment (abhilashshetty04) [reply control-plane/grpc/proto/v1/volume/volume.proto:325]: Done

File: control-plane/grpc/proto/v1/ha/node_agent.proto (hunk 23 above)
Comment (abhilashshetty04) [reply control-plane/grpc/proto/v1/ha/node_agent.proto]: Done

File: control-plane/agents/src/bin/ha/node/server.rs (hunk 22 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/node/server.rs]: Thanks. Didnt know how to filter in map's closure. Incorporated

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 19 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/volume/operations.rs]: Incorporated

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 20 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/volume/operations.rs]: Done

**Review by tiagolobocastro — APPROVED** [review]

**Comment by abhilashshetty04** [comment]
bors merge

**Comment by abhilashshetty04** [comment]
bors cancel

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors merge
